drivers/can/mcp2515.c

```diff
@@ -0,0 +1,748 @@
+
+		/* check that no new interrupts happened while clearing known ones */
+		mcp2515_read_reg(dev, MCP2515_ADDR_CANINTF, &canintf, 1);
+	}
+}
+
+static void mcp2515_int_thread(struct device *dev)
+{
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+
+	while (1) {
+		k_sem_take(&dev_data->int_sem, K_FOREVER);
+		mcp2515_handle_interrupts(dev);
+	}
+}
+
+static void mcp2515_int_gpio_callback(struct device *dev,
+				      struct gpio_callback *cb,
+				      u32_t pins)
+{
+	struct mcp2515_data *dev_data =
+		CONTAINER_OF(cb, struct mcp2515_data, int_gpio_cb);
+
+	k_sem_give(&dev_data->int_sem);
+}
+
+static const struct can_driver_api can_api_funcs = {
+	.configure = mcp2515_configure,
+	.send = mcp2515_send,
+	.attach_msgq = mcp2515_attach_msgq,
+	.attach_isr = mcp2515_attach_isr,
+	.detach = mcp2515_detach
+};
+
+
+static int mcp2515_init(struct device *dev)
+{
+	const struct mcp2515_config *dev_cfg = DEV_CFG(dev);
+	struct mcp2515_data *dev_data = DEV_DATA(dev);
+
+	/* SPI config */
+	dev_data->spi_cfg.operation = SPI_WORD_SET(8);
+	dev_data->spi_cfg.frequency = dev_cfg->spi_freq;
+	dev_data->spi_cfg.slave = dev_cfg->spi_slave;
+
+	dev_data->spi = device_get_binding(dev_cfg->spi_port);
+	if (!dev_data->spi) {
+		SYS_LOG_ERR("SPI master port %s not found", dev_cfg->spi_port);
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_CAN_MCP2515_GPIO_SPI_CS
+	dev_data->spi_cs_ctrl.gpio_dev = device_get_binding(dev_cfg->spi_cs_port);
+	if (!dev_data->spi_cs_ctrl.gpio_dev) {
+		SYS_LOG_ERR("Unable to get GPIO SPI CS device");
+		return -ENODEV;
+	}
+
+	dev_data->spi_cs_ctrl.gpio_pin = dev_cfg->spi_cs_pin;
+	dev_data->spi_cs_ctrl.delay = 0;
+
+	dev_data->spi_cfg.cs = &dev_data->spi_cs_ctrl;
+#else
+	dev_data->spi_cfg.cs = NULL;
+#endif  /* CAN_MCP2515_GPIO_SPI_CS */
+
+	/* Reset MCP2515 */
+	if (mcp2515_soft_reset(dev)) {
+		SYS_LOG_ERR("Soft-reset failed");
+		return -EIO;
+	}
+
+	/* Initialize interrupt handling  */
+	dev_data->int_gpio = device_get_binding(dev_cfg->int_port);
+	if (dev_data->int_gpio == NULL) {
+		SYS_LOG_ERR("GPIO port %s not found", dev_cfg->int_port);
+		return -EINVAL;
+	}
+
+	if (gpio_pin_configure(dev_data->int_gpio, dev_cfg->int_pin,
+			       (GPIO_DIR_IN | GPIO_INT | GPIO_INT_EDGE
+				| GPIO_INT_ACTIVE_LOW | GPIO_INT_DEBOUNCE))) {
+		SYS_LOG_ERR("Unable to configure GPIO pin %u",
+			    dev_cfg->int_pin);
+		return -EINVAL;
+	}
+
+	gpio_init_callback(&(dev_data->int_gpio_cb), mcp2515_int_gpio_callback,
+			   BIT(dev_cfg->int_pin));
+	if (gpio_add_callback(dev_data->int_gpio, &(dev_data->int_gpio_cb))) {
+		return -EINVAL;
+	}
+	if (gpio_pin_enable_callback(dev_data->int_gpio, dev_cfg->int_pin)) {
+		return -EINVAL;
+	}
+	k_thread_create(&dev_data->int_thread, dev_data->int_thread_stack,
+			dev_cfg->int_thread_stack_size,
+			(k_thread_entry_t) mcp2515_int_thread, (void *)dev, NULL, NULL,
+			K_PRIO_COOP(dev_cfg->int_thread_priority), 0, K_NO_WAIT);
+
+	/* Initialize TX data */
+	k_sem_init(&dev_data->tx_sem, 3, 3);
```

Comment:
For all of these sem/mutex maybe you could try using K_SEM_INITIALIZER()/K_MUTEX_INITIALIZER(). It will generate a little less code then. (and it's faster to initialize).

Reply:
These macros are flagged internal, does that mean I can use them in kernel drivers or is that just for really core components? Use throughout the rest of the tree seems to lean towards k_sem_init/k_mutex_init style initialization.

I more dislike the general amount of mutexes and semaphores, but couldn't come up with a smaller amount

Reply:
It's ok to use these macros in native drivers, not in application.

Reply:
Ok, I changed the initializers to be in the mcp2515_data_1 initializer, thanks!